marlowe-chain-sync/chain-indexer/Language/Marlowe/Runtime/ChainIndexer/Database/PostgreSQL.hs

```diff
@@ -567,7 +567,13 @@ commitBlocks = CommitBlocks \blocks ->
       body@(TxBody TxBodyContent{..}) -> case txScriptValidity of
         TxScriptValidity _ ScriptInvalid -> case txReturnCollateral of
           TxReturnCollateralNone     -> []
-          TxReturnCollateral _ txOut -> [SomeTxOut (getTxId body) (TxIx 1) slotNo txOut (Nothing, Nothing) True era]
+          TxReturnCollateral _ txOut -> let
+                                          -- The index of the `TxOut` for collateral change on a transaction that
+                                          -- fails due to an invalid script is one more than the number of `TxOut`s
+                                          -- that there would have been had the transaction succeeded.
+                                          index =  toEnum $ length txOuts
+                                        in
+                                          [SomeTxOut (getTxId body) (TxIx index) slotNo txOut (Nothing, Nothing) True era]
```

Comment:
That's what I thought the spec said :smile:

Reply:
Yes, I shouldn't have committed the previous attempt to fix unless I full understood the spec.

> Change looks good, could you please include a sqitch migration that fixes existing data?

Sadly, the database doesn't contain sufficient information to fix the index. (There is no record of the outputs that would have occurred if the script had succeeded.) This affected 0 records on `preprod` and 2 records on `preview`. I don't know if it affected `mainnet`. Should I add an `update` with a `where` for the two `preview` records to sqitch, since there is no general solution.

Reply:
Oh, right I forgot we don't commit the normal tx outs as well as the collateral returns... then never mind, the databases just need to be wiped and unsynchronized. There may be some cases where we want to do this in a migration, not sure if we want to here though...